algorithms/kernel/svm/svm_train_thunder_workset.h

```diff
@@ -0,0 +1,201 @@
+        _wsIndices.reset(_nWS);
+        DAAL_CHECK_MALLOC(_wsIndices.get());
+
+        return status;
+    }
+
+    size_t getSize() const { return _nWS; }
+
+    services::Status copyLastToFirst()
+    {
+        DAAL_ITTNOTIFY_SCOPED_TASK(select.copyLastToFirst);
+
+        services::Status status;
+        const size_t q = _nWS / 2;
+
+        services::internal::daal_memcpy_s(_wsIndices.get(), q * sizeof(IndexType), _wsIndices.get() + _nWS - q, q * sizeof(IndexType));
+        _nSelected = q;
+        services::internal::service_memset_seq<bool, cpu>(_indicator.get(), false, _nVectors);
+        for (size_t i = 0; i < q; i++)
+        {
+            _indicator[_wsIndices[i]] = true;
+        }
+
+        return status;
+    }
+
+    services::Status select(const algorithmFPType * y, const algorithmFPType * alpha, const algorithmFPType * f, const algorithmFPType C)
+    {
+        DAAL_ITTNOTIFY_SCOPED_TASK(select);
+        services::Status status;
+        IdxValType * sortedFIndices = _sortedFIndices.get();
+
+        for (size_t i = 0; i < _nVectors; ++i)
+        {
+            _sortedFIndices[i].key = f[i];
+            _sortedFIndices[i].val = i;
+        }
+
+        daal::algorithms::internal::qSortByKey<IdxValType, cpu>(_nVectors, sortedFIndices);
+
+        IndexType pLeft  = 0;
+        IndexType pRight = _nVectors - 1;
+        while (_nSelected < _nWS)
+        {
+            if (pLeft < _nVectors)
+            {
+                IndexType i = sortedFIndices[pLeft].val;
+                while (_indicator[i] || !HelperTrainSVM<algorithmFPType, cpu>::isUpper(y[i], alpha[i], C))
+                {
+                    pLeft++;
+                    if (pLeft == _nVectors)
+                    {
+                        break;
+                    }
+                    i = sortedFIndices[pLeft].val;
+                }
+                if (pLeft < _nVectors)
+                {
+                    _wsIndices[_nSelected] = i;
+                    _indicator[i]          = true;
+                    _nSelected++;
+                }
+            }
+
+            if (pRight >= 0)
+            {
+                IndexType i = sortedFIndices[pRight].val;
+                while (_indicator[i] || !HelperTrainSVM<algorithmFPType, cpu>::isLower(y[i], alpha[i], C))
+                {
+                    pRight--;
+                    if (pRight == -1)
+                    {
+                        break;
+                    }
+                    i = sortedFIndices[pRight].val;
+                }
+                if (pRight >= 0)
+                {
+                    _wsIndices[_nSelected] = i;
+                    _indicator[i]          = true;
+                    _nSelected++;
+                }
+            }
+        }
+
+        DAAL_ASSERT(_nSelected == _nWS);
+
+        _nSelected = 0;
+        return status;
+    }
+
+    const IndexType * getIndices() const { return _wsIndices.get(); }
+
+protected:
+    size_t maxpow2(size_t n)
```

Comment:
Consider better function name to avoid confusion

Reply:
Applied